File: python/phone_number.py
import re


class PhoneNumber:

    def __init__(self, number):
        if re.search(r'[a-zA-Z]', number):
            raise ValueError('letters not permitted')

        if re.search(r'[@:!]', number):
            raise ValueError('punctuations not permitted')

        number = re.sub(r'[^0-9]', '', number)

        if len(number) > 11:
            raise ValueError('must not be greater than 11 digits')

        if len(number) == 11:
            if number[0] != '1':
                raise ValueError('11 digits must start with 1')

            number = number[1:]

        if len(number) != 10:
            raise ValueError('must not be fewer than 10 digits')

        if (number[0] == '0'):
            raise ValueError('area code cannot start with zero')
        if (number[0] == '1'):
            raise ValueError('area code cannot start with one')

        self.number = number

        self.area_code = number[:3]

        self.exchange_code = number[3:6]
        if self.exchange_code[0] == '0':
            raise ValueError('exchange code cannot start with zero')
        if self.exchange_code[0] == '1':
            raise ValueError('exchange code cannot start with one')

        self.subscriber_number = number[6:]

    def pretty(self):
        return f'({self.area_code})-{self.exchange_code}-{self.subscriber_number}'

File: elixir/phone_number.ex
defmodule PhoneNumber do
  @doc """
  Remove formatting from a phone number if the given number is valid. Return an error otherwise.
  """
  @spec clean(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def clean(raw) do
    s = String.replace(raw, ~r/[^0-9]/, "")
    len = String.length(s)

    cond do
      Regex.match?(~r/[^0-9()-. ]/, raw) -> {:error, "must contain digits only"}
      len < 10 -> {:error, "must not be fewer than 10 digits"}
      len == 11 and String.starts_with?(s, "1") -> String.slice(s, 1, 10) |> do_clean()
      len == 11 -> {:error, "11 digits must start with 1"}
      len > 11 -> {:error, "must not be greater than 11 digits"}
      true -> do_clean(s)
    end
  end

  defp do_clean(s) do
    cond do
      Regex.match?(~r/^0/, s) -> {:error, "area code cannot start with zero"}
      Regex.match?(~r/^1/, s) -> {:error, "area code cannot start with one"}
      Regex.match?(~r/^...0/, s) -> {:error, "exchange code cannot start with zero"}
      Regex.match?(~r/^...1/, s) -> {:error, "exchange code cannot start with one"}
      true -> {:ok, s}
    end
  end
end
